File: Java/Data Structures & Algorithms Module - Java/Week8/Revision/Week7/CleanString.java
/*
CleanString.java
Billy Cussen
17/11/2020
*/

public class CleanString{

    public static void main(String [] args){

        System.out.println(cleanString("aaabbbbxxxxdddddwwwwww"));
        System.out.println(cleanStringRecursion("aaabbbbxxxxdddddwwwwww"));
    }

    public static String cleanString(String word){
        StringBuffer sb = new StringBuffer();
        char a = word.charAt(0);
        sb.append(a);
        for(int i = 1; i < word.length();i++){
            if(a==word.charAt(i)){
                sb.append("");
            } else{
                a = word.charAt(i);
                sb.append(a);
            }
        }
        return sb.toString();
    }

    public static String cleanStringRecursion(String word){
        if(word.length()<=1){
            return word;
        }

        if(word.charAt(0)==word.charAt(1)){
            return "" + cleanStringRecursion(word.substring(1,word.length()));
        } else{
            return word.charAt(0) + cleanStringRecursion(word.substring(1,word.length()));
        }
    }

}

File: Python/Data Structures & Algorithms Module - Python/Week7/CleanString.py
"""
CleanString.py
Billy Cussen
17/11/2020
"""

def cleanString(word):
    res = ""
    s = set()
    for c in word:
        if c not in s:
            res+=c
            s.add(c)
    return res

def cleanStringRecursion(word):
    if len(word) == 1:
        return word

    if word[1] == word[0]:
        return cleanStringRecursion(word[1:])
    else:
        return word[0]+cleanStringRecursion(word[1:])

print(cleanString("aaabbbccccdddddd"))
print(cleanStringRecursion("aaaaabbbbbbccccccddddddeeeeeeffffggggg"))